### PR TITLE
WiP: Return 405 Method not Allowed

### PR DESCRIPTION
**Pull request by omgitsbillryan**
When a request comes to `vets-api` where the path _is_ matched but the HTTP verb _is not_, then we want to:

- Return code `405`
- Return a `Allowed` header that contains a CSV of allowed verbs (e.g. `Allowed : GET, POST`)

### CODE REVIEW

**Comment by omgitsbillryan** [comment]
I'm struggling with writing an rspec for this case.  This is what I have tried (and failed) inside `application_controller_spec`:
```ruby
controller do
  def some_route
    # Where I end up
  end

  def unmatched
    # Where I *want* to end up
  end
end

context 'when app globs unmatched routes' do
  before(:each){ routes.draw { 
    get 'some_route', to: 'anonymous#some_route'
    match '*path', to: 'anonymous#unmatched', via: %i(get post put patch delete) 
    }
   }
  it 'returns a 405 when a path is matched but the verb does not' do
    post :some_route         # lands me in some_route
    get '/path/doesnt/exist' # spits an error out
  end
end
```

Maybe there's a better way to go about testing this case with an rspec

**Comment by omgitsbillryan** [comment]
@saneshark i think this might be up your alley - any pointers you could give me on how to write an rspec for this?

File: lib/common/exceptions/method_not_allowed.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+module Common
```
Comment (saneshark) [inline lib/common/exceptions/method_not_allowed.rb]: currently all of these common exceptions have a unique spec file. Would you mind adding one for this PR. Can copy and paste most of it -- I will likely refactor in the future into "shared examples"

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -16,7 +16,11 @@ def cors_preflight
   end
 
   def routing_error
-    raise Common::Exceptions::RoutingError, params[:path]
+    if allowed_verbs&.empty?
+      raise Common::Exceptions::RoutingError, params[:path]
+    else
+      raise Common::Exceptions::MethodNotAllowed, allowed_verbs
+    end
   end
```
Comment (omgitsbillryan) [inline app/controllers/application_controller.rb]: Just by looking at it - wont it just result in a 404? (from line 29)
```ruby
#   raise Common::Exceptions::RoutingError
```

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -112,4 +118,16 @@ def pagination_params
   def render_job_id(jid)
     render json: { job_id: jid }, status: 202
   end
+
+  def allowed_verbs
+    all_methods = ActionDispatch::Routing::HTTP_METHODS
```
Comment (saneshark) [inline app/controllers/application_controller.rb]: consider using this directly in the call below or maybe using a private method, doesn't appear as the local variable is used more than once, here.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -112,4 +118,16 @@ def pagination_params
...
+    @allowed_verbs ||= all_methods.select do |verb|
+      begin
+        match = Rails.application.routes.recognize_path(params[:path], method: verb)
+        match[:action] != 'routing_error' && match[:action] != 'cors_preflight'
+      rescue ActionController::RoutingError
+        nil
```
Comment (saneshark) [inline app/controllers/application_controller.rb]: shouldn't routing error render #routing_error

**Comment by saneshark** [comment]
@omgitsbillryan take a look at what I had to do in spec/controllers/application_controller_spec.rb :)

It's super weird but it gets the job done.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (saneshark) [inline app/controllers/application_controller.rb]: You can change that to raise the Exception you created.

**Comment by saneshark** [comment]
@omgitsbillryan actually, now that I think about it, this will probably be more helpful for you:

https://github.com/department-of-veterans-affairs/vets-api/blob/master/spec/request/exceptions_request_spec.rb

Just use a GET for a method only implements POST or vice-versa.

**Comment by omgitsbillryan** [comment]
Closing for now - this PR introduces too much risk for too little reward.

**Pull request closed without merging**